// generator-arc/generators/store/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _yeomanGenerator = require('yeoman-generator');

var _inquirerAutocompletePrompt = require('inquirer-autocomplete-prompt');

var _inquirerAutocompletePrompt2 = _interopRequireDefault(_inquirerAutocompletePrompt);

var _utils = require('../utils');

var _helpers = require('./helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _class = function (_Base) {
  _inherits(_class, _Base);

  function _class() {
    var _ref;

    _classCallCheck(this, _class);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = _possibleConstructorReturn(this, (_ref = _class.__proto__ || Object.getPrototypeOf(_class)).call.apply(_ref, [this].concat(args)));

    _this.getStores = function () {
      var _this$options = _this.options,
          ours = _this$options.ours,
          theirs = _this$options.theirs;

      var all = !ours && !theirs;
      var stores = [];
      if (all || ours) {
        stores.push.apply(stores, _toConsumableArray((0, _helpers.findStorePaths)(process.cwd()).map(_helpers.getInfoFromStorePath)));
      }
      if (all || theirs) {
        stores.push.apply(stores, _toConsumableArray((0, _helpers.findStorePaths)((0, _utils.defaultTemplatePath)()).map(function (path) {
          return (0, _helpers.getInfoFromStorePath)(path, (0, _utils.branchUrl)());
        })));
      }
      return stores;
    };

    _this.env.adapter.promptModule.registerPrompt('autocomplete', _inquirerAutocompletePrompt2.default);
    _this.option('ours');
    _this.option('theirs');
    _this.stores = _this.getStores();
    return _this;
  }

  _createClass(_class, [{
    key: 'prompting',
    value: function prompting() {
      var _this2 = this;

      var choices = this.stores.map(function (store) {
        return {
          name: store.name + ' ' + _chalk2.default.gray(store.url),
          short: store.name,
          value: store
        };
      });

      var prompts = [{
        type: 'autocomplete',
        name: 'store',
        message: 'Which store do you want to clone?',
        source: /* istanbul ignore next */function source(answers, input) {
          return Promise.resolve(input ? choices.filter(function (choice) {
            return choice.name.indexOf(input) >= 0;
          }) : choices);
        }
      }, {
        type: 'input',
        name: 'name',
        message: function message(_ref2) {
          var store = _ref2.store;
          return 'How do you want to name the ' + store.name + ' store?';
        },
        default: function _default(_ref3) {
          var store = _ref3.store;
          return store.name;
        }
      }];

      return this.prompt(prompts).then(function (answers) {
        _this2.answers = answers;
      });
    }
  }, {
    key: 'writing',
    value: function writing() {
      var _this3 = this;

      var _answers = this.answers,
          name = _answers.name,
          store = _answers.store;

      var isTheirs = store.url.indexOf('https') === 0;
      var names = (0, _helpers.transformName)(name);
      var originalNames = (0, _helpers.transformName)(store.name);
      var dPath = function dPath() {
        return _this3.destinationPath(_helpers.getStorePath.apply(undefined, arguments));
      };
      var tPath = function tPath() {
        return isTheirs ? (0, _utils.defaultTemplatePath)(_helpers.getStorePath.apply(undefined, arguments)) : dPath.apply(undefined, arguments);
      };

      _fs2.default.readdirSync(tPath(store.name)).forEach(function (file) {
        var contents = (0, _helpers.replaceNames)(_this3.fs.read(tPath(store.name, file)), originalNames, names);
        _this3.fs.copy(tPath(store.name, file), dPath(name, file));
        _this3.fs.write(dPath(name, file), contents);
      });
    }
  }]);

  return _class;
}(_yeomanGenerator.Base);

exports.default = _class;
module.exports = exports['default'];
